server/src/main/java/org/elasticsearch/search/aggregations/metrics/AbstractHDRPercentilesAggregator.java

```diff
@@ -106,6 +118,36 @@ public void collect(int doc, long bucket) throws IOException {
         };
     }
 
+    private LeafBucketCollector collectHistogramValues(final HistogramValues values,
+                                                       final BigArrays bigArrays, final LeafBucketCollector sub) {
+        return new LeafBucketCollectorBase(sub, values) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                states = bigArrays.grow(states, bucket + 1);
+                DoubleHistogram state = states.get(bucket);
+                if (state == null) {
+                    state = new DoubleHistogram(numberOfSignificantValueDigits);
+                    // Set the histogram to autosize so it can resize itself as
+                    // the data range increases. Resize operations should be
+                    // rare as the histogram buckets are exponential (on the top
+                    // level). In the future we could expose the range as an
+                    // option on the request so the histogram can be fixed at
+                    // initialisation and doesn't need resizing.
+                    state.setAutoResize(true);
+                    states.set(bucket, state);
+                }
```

Comment:
can we extract the logic of creating a new histogram to a separate method that is used by both collectNumeric and collectHistogramValues?

Reply:
done